dynamically use free port for callback service?

If I have a notebook with rikai running, i get an error when i also try to run the unit tests

```
python/rikai/spark/sql/__init__.py:37: in init
    init_cb_service(spark)
python/rikai/spark/sql/callback_service.py:31: in init_cb_service
    jvm.start_callback_server(callback_server_parameters=params)
../../../miniconda3/envs/eto/lib/python3.7/site-packages/py4j/java_gateway.py:1894: in start_callback_server
    self._callback_server.start()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <py4j.java_gateway.CallbackServer object at 0x7f8eca99fdd0>

    def start(self):
        """Starts the CallbackServer. This method should be called by the
        client instead of run()."""
        af_type = socket.getaddrinfo(self.address, self.port)[0][0]
        self.server_socket = socket.socket(af_type, socket.SOCK_STREAM)
        set_reuse_address(self.server_socket)
        try:
            self.server_socket.bind((self.address, self.port))
            # 4-tuple for ipv6, 2-tuple for ipv4
            info = self.server_socket.getsockname()
            self._listening_address = info[0]
            self._listening_port = info[1]
        except Exception as e:
            msg = "An error occurred while trying to start the callback "\
                  "server ({0}:{1})".format(self.address, self.port)
            logger.exception(msg)
>           raise Py4JNetworkError(msg, e)
E           py4j.protocol.Py4JNetworkError: An error occurred while trying to start the callback server (127.0.0.1:25334)

../../../miniconda3/envs/eto/lib/python3.7/site-packages/py4j/java_gateway.py:2216: Py4JNetworkError

```

--- a/python/rikai/spark/sql/callback_service.py
+++ b/python/rikai/spark/sql/callback_service.py
@@ -30,7 +30,14 @@ def init_cb_service(spark: SparkSession):
         # Re-use the auth-token from the main java/spark process
         auth_token=jvm.gateway_parameters.auth_token,
     )
+
     jvm.start_callback_server(callback_server_parameters=params)
+
+    python_port = jvm.get_callback_server().get_listening_port()
+    jvm.java_gateway_server.resetCallbackClient(
+        jvm.java_gateway_server.getCallbackClient().getAddress(),
+        python_port,
+    )
     logger.info("Spark callback server started")
 
     cb = CallbackService(spark)

--- a/python/rikai/spark/sql/callback_service.py
+++ b/python/rikai/spark/sql/callback_service.py
@@ -25,6 +25,8 @@ def init_cb_service(spark: SparkSession):
     params = CallbackServerParameters(
         daemonize=True,
         daemonize_connections=True,
+        # https://www.py4j.org/advanced_topics.html#using-py4j-without-pre-determined-ports-dynamic-port-number
+        port=0,
         # Re-use the auth-token from the main java/spark process
         auth_token=jvm.gateway_parameters.auth_token,
     )

